src/core/storage.rs

```diff
@@ -158,11 +162,20 @@ impl<K: Hash + Clone + Eq> RecordsStorage<K> {
 		self.ephemeral_records
 			.entry(block_number)
 			.or_insert_with(Default::default)
-			.insert(key, entry);
+			.insert(key, Arc::downgrade(&entry));
 
 		self.prune();
 	}
 
+	pub fn replace(&mut self, key: K, entry: StorageEntry) -> Option<StorageEntry> {
```

Comment:
We could merge this into the insert() method.

Reply:
Well, that might not be a good idea I think, as these methods have different signatures and meaning. Probably, we could use something like `entry` method for `HashMap` but it will introduce complexity and since we use storage for storing unique and distinct hashes, we have literally two different cases: 1) insert a new unique key (maybe even panic on duplicate) or 2) replace/update definitely or likely known key.

Reply:
Yeah, that makes sense only if we want to have uniqueness semantics for the write API.